packages/eslint-plugin/docs/rules/no-extraneous-class.md

```diff
@@ -2,14 +2,21 @@
 
 Forbids the use of classes as namespaces.
 
-This rule warns when a class is accidentally used as a namespace.
+This rule warns when a class has no non-static members, such as for a class used exclusively as a static namespace.
 
 ## Rule Details
 
-From TSLint’s docs:
+Users who come from a [OOP](https://en.wikipedia.org/wiki/Object-oriented_programming) paradigm may wrap their utility functions in an extra class, instead of putting them at the top level of an ECMAScript module.
+Doing so is generally unnecessary in JavaScript and TypeScript projects.
 
-> Users who come from a Java-style OO language may wrap their utility functions in an extra class,
-> instead of putting them at the top level.
+- Wrapper classes add extra runtime bloat and cognitive complexity to code without adding any structural improvements
```

Comment:
@goktugyil https://github.com/typescript-eslint/typescript-eslint/issues/5040#issuecomment-1134013764:

> Do we have any data on the performance difference? Is it 5% difference? 0.01% difference?

I don't have hard data to draw from, as relatively fewer TypeScript projects nowadays still use classes like static namespaces. IIRCC @ian-craig and/or @JFrei86 may have some from their times investigating performance in Microsoft Office (which was at least attempting to move away from namespaces a year or two back).

But, just looking at the difference between these two JavaScript snippets:

```ts
class Settings {
    static version = '1.2.3';
}

console.log(Settings.version);
```

```ts
const version = '1.2.3';

console.log(version);
```

We can see that the class version adds an extra `Settings.` lookup in both space (code size) and time (runtime lookup) to each call. This is going to be at the very least slightly negative, especially before JIT kicks in.

Reply:
Classes are slower than objects to define at runtime!
A quick benchmark shows they're around 99% slower to define in chrome: https://jsbench.me/93l3j9tjxc/1

Ignoring the definition, they're exactly as performant as one-another to just call:
https://jsbench.me/vyl3j9vx52/1

Reply:
@JoshuaKGoldberg I think in this context, it would be more appropriate to compare `namepace.version` to `staticClass.version`

Reply:
<img width="752" alt="Screen Shot 2022-05-23 at 18 37 03" src="https://user-images.githubusercontent.com/5816878/169915753-c241a6b5-da3f-4343-9254-50eb32a140a2.png">

Just adding a little bit of logic seems to change it a lot. So yes maybe 99% slower in isolation. But practically speaking, I don't see a performance difference between 2M ops/s vs 364M ops/s

As a project would need a minimum 20K static classes to even begin any performance issues caused by this. Right?

Reply:
Yeah definitely - sorry I should have clarified the minute scale you're talking about for declarations. Whilst it's 99% faster - you're talking about `0.0000005`s per declaration vs `0.000000003`s per declaration. Sure it's slower, but you won't ever notice the difference in practice.

That's why I linked the method call case separately - ultimately they're exactly the same perf in practice.

There may be memory differences - but no doubt you'd be talking in practice about such a minute difference that it makes no practical difference.

Reply:
One thing I am curious is, the define time it take only happens when it is first loading the javascript right? After the first load, they're the same on running?

Reply:
> One thing I am curious is, the define time it take only happens when it is first loading the javascript right?

yeah define time happens when the thing is first parsed and defined - so exactly once per thing.

Then the member access is pretty well a constant-time, regardless of whether it's a class or object.

Reply:
I think we can probably close this thread with a resolution to remove _"runtime bloat"_ from the docs. Although it's technically true, the performance aspects are almost certainly not a concern for the vast majority of applications. The _cognitive complexity_ point is more important.

Reply:
I disagree about removing it.

- It is a correct argument? It is.
- Does it weaken the rest of the arguments against classes? No.
- Does it add motivation to not using classes? Yes, because there are a non-trivial number of people in favor of micro-optimizations, especially when such optimizations come with addition readability improvements (which is rare!)
- Have we considered how Babel/tsc is transpiling classes? Transpiled classes can be significantly larger in size and slower at execution.

Reply:
IMO it's really *not* a correct argument because we're talking about a difference of 400 ***nano** seconds (`4x10^-7`s) - **at definition type only** (i.e. once per "run").

Put another way to make it clear why it's not a correct argument - if your codebase had 1,000,000 utility classes and you converted them all to objects - you would save around 0.4 seconds of initialisation time. That's it.
A more sane upper-bound number would be, say, 1,000 utility classes - which would only save 0.0004 seconds - a time scales which is literally imperceptible to a user!

I definitely agree that cognitive complexity, standardisation, etc are all valid reasons! I strongly believe in the rule and the code standard it enforces.
But it's disingenuous for us to represent it as runtime bloat or a performance optimisation, and thus that argument specifically does not warrant being mentioned at all.

> Have we considered how Babel/tsc is transpiling classes? Transpiled classes can be significantly larger in size and slower at execution.

I think you'd find the opposite - old-school classes utilise functions which are very long-lived language features, and are thus well optimised.

I used the TS playground to transpile my benchmark before, and I got the following results for a definition:
- static class: 2,325,851 ops/s
- transpiled static class: 208,966,313 ops/s
- object: 367,688,543 ops/s

https://jsbench.me/93l3j9tjxc/2

So (in chrome) using a transpiled class is actually better than using a native class!
And the difference in performance between an object and a transpiled class is even smaller (2 nano seconds (`2x10^-9`s))!

Additional testing numbers:
1) in chrome - native = 99.3% slower, transpiled = 42.1% slower
1) in safari - native = 89.2% slower, transpiled = 37.3% slower
1) in firefox - native = 99.9% slower, transpiled = 95.5% slower
   - firefox has super-optimised objects from the looks of it at over 1.5b ops/s

Again - performance is not an argument we should include, but everything else 100% has merit!